Implement RVU Genie client device discovery

ehendrix23 opened the pull request:
## Description:

Perform discovery of RVU Genie client devices allowing them to be discovered even when they are offline upon discovery of the main Genie DVR.

**Pull request in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) with documentation (if applicable):** home-assistant/home-assistant.io#7768

## Example entry for `configuration.yaml` (if applicable):
```yaml
media_player:
  - platform: directv
    host: 192.168.1.2
    port: 8080
    name: Family_Room
    device: 0
    discover_clients: true
    client_discover_interval: 20
  - platform: directv
    host: 192.168.1.2
    port: 8080
    name: Master Bedroom
    device: 0
    discover_clients: false
```

## Checklist:
  - [X] The code change is tested and works locally.
  - [X] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [X] There is no commented out code in this PR.

If user exposed functionality or configuration variables are added/changed:
  - [X] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [X] New or updated dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.

If the code does not interact with devices:
  - [X] Tests have been added to verify that the new code works.

ehendrix23 committed:
Updated requirements

Updated requirements to directpy 0.6

balloob commented on homeassistant/components/media_player/directv.py:
```diff
@@ -47,73 +55,287 @@
     vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
     vol.Optional(CONF_PORT, default=DEFAULT_PORT): cv.port,
     vol.Optional(CONF_DEVICE, default=DEFAULT_DEVICE): cv.string,
+    vol.Optional(CONF_DISCOVER_CLIENTS, default=DEFAULT_DISCOVER_CLIENTS):
```
No need to make this or discover interval configurable.

balloob commented on homeassistant/components/media_player/directv.py:
```diff
@@ -47,73 +55,287 @@
+
+        client_discover_interval = config.get(CONF_CLIENT_DISCOVER_INTERVAL,
+                                              DEFAULT_CLIENT_DISCOVER_INTERVAL)
+        if client_discover_interval < timedelta(seconds=5):
```
Drop the confguration, also validation should happen in the config schema.

balloob commented on homeassistant/components/media_player/directv.py:
```diff
@@ -47,73 +55,287 @@
+            CONF_HOST: host,
+            CONF_PORT: port,
+            CONF_DEVICE: device,
+            RECEIVER_ID: resp.get('receiverId', host).replace(' ', ''),
```
your default values are `[]` which doesn't have a `.get`

balloob commented on homeassistant/components/media_player/directv.py:
```diff
@@ -47,73 +55,287 @@
+        serial_num = discovery_info.get('serial', host)
+        receiver_id = discovery_info.get('serial', '').split('-')[-1]
+        receiver_id = receiver_id if receiver_id else host
+        name = 'DirecTV_{}'.format(serial_num)
```
no underscores in names.

balloob commented on homeassistant/components/media_player/directv.py:
```diff
@@ -47,73 +55,287 @@
+        name = 'DirecTV_{}'.format(serial_num)
+        resp = []
+
+        if (receiver_id, DEFAULT_DEVICE) in known_devices:
```
This will search if the tuple `(receiver_id, DEFAULT_DEVICE)` exists inside `known_devices`, is that what you want?

ehendrix23 replied:
Yes, that is exactly what I want. I'm checking if that receiver was already added as an entity. A receiver is unique through its combination of receiver_id (serial number) and device ID.

balloob commented on homeassistant/components/media_player/directv.py:
```diff
@@ -47,73 +55,287 @@
-            }
-
-        _LOGGER.debug("Known devices: %s", known_devices)
+            _LOGGER.debug("Request exception %s trying to connect to %s",
```
This seems very similar to the code above? Can you consolidate?

balloob commented on homeassistant/components/media_player/directv.py:
```diff
@@ -47,73 +55,287 @@
+            CONF_HOST: host,
+            CONF_PORT: DEFAULT_PORT,
+            CONF_DEVICE: DEFAULT_DEVICE,
+            RECEIVER_ID: receiver_id,
```
In the other case you replaced spaces with empty strings for receiver id.

ehendrix23 replied:
The difference is how the serial number is received. When the serial number is retrieved through the API then it comes in as 3 groups of 4 numbers each separated by a space: 1234 5678 9012
Through the HASS discovery, we also get the serial number but then it comes in as: DirecTV-123456789012

Hence, when getting it through the API we remove the spaces resulting in: 123456789012
But when serial is from the discovery then instead we split on - and grab the second (or last) entry resulting in: 123456789012

balloob commented on homeassistant/components/media_player/directv.py:
```diff
@@ -47,73 +55,287 @@
+                DEFAULT_CLIENT_DISCOVER_INTERVAL,
+        }
+
+    if directv_entity is not None:
```
guard clause

```python
if directv_entity is None:
    return
```

balloob commented on homeassistant/components/media_player/directv.py:
```diff
@@ -47,73 +55,287 @@
+
+    def _discover_directv_client_devices(self, now=None):
+        """Discover new client devices connected to the main DVR."""
+        known_devices = self._hass.data.get(DATA_DIRECTV, set())
```
Just call `setdefault` once at the start of setup platform and you don't have to do this anymore.

ehendrix23 replied:
And I was already using setdefault :-)

balloob commented on homeassistant/components/media_player/directv.py:
```diff
@@ -53,67 +57,244 @@
+                receiver_id_resp = None
+
+            if not receiver_id_resp:
+                self._receiver_id = receiver_id_resp.get('receiverId')
```
This will crash, because you check if `receiver_id_resp` is falsey, and then try to get data from it?

ehendrix23 committed:
Updated based on feedback

ehendrix23 committed:
Removed abstract classes

Removed again, thought this was causing issue with pylint with seems plint has an issue with latest version of aiohttp and Python 3.7

rohankapoorcom commented on homeassistant/components/media_player/directv.py:
```diff
@@ -54,67 +58,241 @@
 
+    discovered = False
     if CONF_HOST in config:
+        name = config.get(CONF_NAME)
```
Use [] syntax for anything that is required or an optional with a default.

Example: `name = config[CONF_NAME]`

andrewsayre commented:
This PR has gone stale.  Please rebase and open a new PR if you are still interested in submitting this change.
